File: felix/state/state.hs
import Control.Monad.State.Strict
import Data.Functor.Identity

data Counter = Counter{value :: Int, numberOfIncrements :: Int} deriving (Show)

incrementCounter :: Counter -> Identity (Int, Counter)
incrementCounter (Counter{value=oldValue, numberOfIncrements=oldNumberOfIncrements}) = 
  Identity (oldValue, Counter{value = succ oldValue, numberOfIncrements = succ oldNumberOfIncrements})

incrementCounterState :: State Counter Int
incrementCounterState = StateT incrementCounter

incrementCounterTwice :: State Counter Int
incrementCounterTwice = incrementCounterState >>= (\_ -> incrementCounterState)

incrementCounterThrice :: State Counter Int
incrementCounterThrice = 
  incrementCounterState >>= 
  \_ -> incrementCounterState >>= 
  \_ -> incrementCounterState

{-
  >:t incrementCounterState
  incrementCounterState :: State Counter Int
  >:t runState incrementCounterState
  runState incrementCounterState :: Counter -> (Int, Counter)


  >let ourCounter=Counter{value=24, numberOfIncrements=3}
  >(runState incrementCounterState) ourCounter
  (24,Counter {value = 25, numberOfIncrements = 4})
-}

{-
  (>>=) :: State s a -> (a -> State s b) -> State s b

  >>= Uses the final state of the first computation as the initial
      state of the second.

  It's implemented like this.

  instance Monad (State s) where
      return a = State $ \s -> (a, s)
      m >>= k = State $ \s -> let (a, s') = runState m s in runState (k a) s'
-}

ret :: a -> State s a
ret x = StateT (\s -> Identity (x, s))

mbind :: State s a -> (a -> State s b) -> State s b
mbind m k = StateT ( \s -> let (a, s') = runState m s in Identity (runState (k a) s') )

incrementCounterTwice' :: State Counter Int
incrementCounterTwice' = mbind incrementCounterState (\_ -> incrementCounterState)

-- *Main Control.Monad.State.Strict> let ourCounter=Counter{value=20, numberOfIncrements=3}
-- *Main Control.Monad.State.Strict> (runState incrementCounterThrice) ourCounter
-- (22,Counter {value = 23, numberOfIncrements = 6})
-- *Main Control.Monad.State.Strict> 

File: ocaml/state/state.ml
(* State monad *)

type ('s, 'a) state =
  | State of ('s -> 'a * 's)
  ;;

let run : ('s, 'a) state -> 's -> ('a * 's) =
  fun x ->
     match x with
       | State u -> u
  ;;

let bind : ('s, 'a) state -> ('a -> ('s, 'b) state) -> ('s, 'b) state = 
  fun m k -> 
    State (
      fun state0 -> 
        let (a, state1) =  run m state0 in run (k a) state1
    )
  ;;

let return_ : 'a -> ('s, 'a) state = 
  fun x ->
    State (fun st -> (x, st))
  ;;
  
let (>>=) : ('s, 'a) state -> ('a -> ('s, 'b) state) -> ('s, 'b) state  = 
  fun m k -> bind m k 
;;

(* Basic test. *)

type counter = { v : int ; n : int ; } ;;

let increment_counter : counter -> (int * counter) = 
  fun s ->
    let value = s.v
    in let s' = {v = succ s.v; n = succ s.n }
       in (value, s')
  ;;

let increment_counter_state : (counter, int) state = 
  State increment_counter
  ;;

let res = 
  run 
    (
       increment_counter_state >>= 
        (fun _ -> increment_counter_state)
    )
    {v = 1; n = 2}
in
    match res with
    | (a, s) -> (Printf.printf "{a = %d, v = %d, n = %d}" a s.v s.n) 
    ;;

